### PR TITLE
Handle redirects for Yun shield sketch upload

### PR DESCRIPTION
**Pull request by sandeepmistry**
When POST'ing a sketch for upload to `/data/upload_sketch_silent`, it returns a 302 Found response with a Location of `/cgi-bin/luci/;stok=<stok>/data/upload_sketch_silent`. The Go HTTP client will attempt to do a GET request for this redirect (instead of POST) and get a 401 response because there is no Authentication header in the request.

This change will make `spProgramNetwork` POST a new request to the redirected URL when this condition is encountered.

It would be nice to change the behaviour of the Yun Shield REST service to not redirect, because the sketch data is sent twice over the network. If this is not possible, then we can use this patch instead.

cc/ @facchinm @matteosuppo

### CODE REVIEW

File: programmer.go [hunk 1]
```diff
@@ -144,69 +145,99 @@ func spProgramNetwork(portname string, boardname string, filePath string, authda
...
+
+		// Submit the request
+		client := &http.Client{}
+		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+			return errors.New("redirect")
+		}
+		res, err := client.Do(req)
+
+		// clear redirect errors and use status code
+		if err != nil && strings.HasSuffix(err.Error(), "redirect") {
```
Comment (matteosuppo) [inline programmer.go]: Shouldn't it return a 302 if it's a redirect? Is it considered an error?

File: programmer.go [hunk 2]
```diff
@@ -144,69 +145,99 @@ func spProgramNetwork(portname string, boardname string, filePath string, authda
...
-		return err
-	}
-	if _, err = fw.Write([]byte(colonToUnderscore(boardname))); err != nil {
-		log.Println("Error writing form field")
-		return err
-	}
-	// Don't forget to close the multipart writer.
-	// If you don't close it, your request will be missing the terminating boundary.
-	w.Close()
+	var cookies []*http.Cookie
```
Comment (matteosuppo) [inline programmer.go]: Why does it need cookies?

File: programmer.go [hunk 3]
```diff
@@ -144,69 +145,99 @@ func spProgramNetwork(portname string, boardname string, filePath string, authda
...
-		log.Println("Error creating post request")
-		return err
-	}
-	// Don't forget to set the content type, this will contain the boundary.
-	req.Header.Set("Content-Type", w.FormDataContentType())
-	if authdata.Username != "" {
-		req.SetBasicAuth(authdata.Username, authdata.Password)
-	}
+		// Now that you have a form, you can submit it to your handler.
+		req, err := http.NewRequest("POST", _url, &b)
```
Comment (matteosuppo) [inline programmer.go]: This thing is indented weirdly: is it a github glitch?

File: programmer.go [hunk 4]
```diff
@@ -144,69 +145,99 @@ func spProgramNetwork(portname string, boardname string, filePath string, authda
...
+		if err != nil {
+			log.Println("Error creating post request")
+			return err
+		}
+		// Don't forget to set the content type, this will contain the boundary.
+		req.Header.Set("Content-Type", w.FormDataContentType())
+		if authdata.Username != "" {
+			req.SetBasicAuth(authdata.Username, authdata.Password)
```
Comment (matteosuppo) [inline programmer.go]: did you launch `go fmt`?

File: programmer.go (hunk 2 above)
Comment (sandeepmistry) [inline programmer.go]: If you don't set a cookie from the 302 response, you keep on redirecting.

We need to discuss this behaviour with @facchinm.

File: programmer.go (hunk 4 above)
Comment (sandeepmistry) [inline programmer.go]: Yes, I think it is just how the diff is displayed. Everything is tabbed in one more level because of the loop for redirect retry.

**Comment by facchinm** [comment]
The only problem I see here is that the Yun101 (or a Yun updated to fw 1.6.2) will not respond a `302` to the `upload_sketch_silent` request, thus avoiding the fallback to ssh method (which would work otherwise)

**Comment by sandeepmistry** [comment]
@facchinm yes, it responds with a 404, and we fallback to SSH mode.

**Comment by zmoog** [comment]
The support for Yun Shield was never completed, so we're closing this PR. RIP.

**Pull request closed without merging**